### ISSUE TITLE
Storage Live Tests infra improvements - stability and scale

### ISSUE THREAD

**Issue by kasobol-msft** [Azure/azure-sdk#2025]
This ask is a follow up from latest attempt to get storage live tests in .NET green. However, it's applicable across all languages where storage sdk live tests are run.

The issues identified:
1. There's desire to test multiple service versions. Today one service version takes ~1hr to test. Times all service versions used to yield 6hr runtime which is not acceptable. The short-term mitigation applied is to test only latest service version in nightly builds (but that created a gap that needs to be addressed).
2. There's hard limit of 250 storage accounts per subscriptions. Each test run requires multiple accounts as some features can be enabled only at creation or some tests require isolation at account level. We've been observing that limit being hit already. Worst case scenario is `number_of_accounts_per_run*number_of_platforms*project_references_vs_package_references*number_of_languages` with more dimensions coming if we attempt to find ways to parallelize more.
3. The amount of tests we have might put high load on single storage account saturating one of it's limits. We've seen blips like this few times. That will become more of a problem if we attempt to more aggressively parallelize tests. (however, this issue seems to be least problematic now).
4. There's small probability that account is created on less stable infrastructure or is created with undesired feature configuration (i.e. lower service version enabled). While this problem is being address it would be great to have ability to automatically re-try failed job before giving up.

We'd like to:
- Decrease runtime (faster tests are better)
- Maintain coverage (we shouldn't address this by disabling tests or service versions)
- Reduce blast radius (single test failure shouldn't derail health of unrelated package, i.e. blob test failure shouldn't mark queues SDK unhealthy) (re-running well scoped test suite is faster than re-running monolith).

Proposed improvements to address these issues:
1. Using separate subscription per language.
2. Adding ability to use multiple subscriptions in one language (i.e. sharding test creation across multiple subscriptions).
3. Splitting tests per storage service.
4. Enable parallelization on reasonable level (test class or test method is fine) including playback/recording mode as that would benefit from faster run time as well (see https://github.com/Azure/azure-sdk-for-net/issues/6905 )
5. Ability to parallelize testing different service version either through separate jobs within pipelines or separate pipelines (to spread the load between agents and storage accounts)
6. Have Smoke Test (few fast tests) that prove generated resources through ARM are healthy and have desired configuration before commencing full run.

**Comment by kurtzeborn** [comment]
@benbp, there's a lot being asked in here.  Please sync with @weshaggard and @mitchdenny before you start trying to tackle this one.

Other work we're doing on the EngSys that might affect the possible solutions to the problems being presented here:
- Transitioning to 1ES scale set machines for Linux and Windows.  This could result in run times improving because the machines have more resources available and perform better.
- The work you're doing on standardizing how we target specific subscriptions from within the pipeline.
- [Proposed work] Integrating the test pipelines into the unified release pipelines.

**Comment by benbp** [comment]
I've created a couple issues to address the higher priority items here @kurtzeborn @kasobol-msft :

- Service version testing and configuration: https://github.com/Azure/azure-sdk/issues/2048 (meta-level support for arbitrary test matrix dimensions and filters).
  - The above issue will also cover a lot of the other usability/test speed/coverage scenarios raised here.
- Multiple subscription sharding: https://github.com/Azure/azure-sdk/issues/2049
- Splitting tests per storage service: https://github.com/Azure/azure-sdk/issues/2050

I've chatted with @weshaggard and @mitchdenny regarding all the items, but it may be easier to have a discussion about this and share notes back to this thread instead.
